### PR TITLE
Make startHook not blocking jenkins startup

### PR DESCRIPTION
**Pull request by bjoernhaeuser**
This PR is two folded. The first commits holds a lot of whitespace fixes. 

The second commit is the actually change. It moves the creating of the hook into a runnable, which makes it way faster on startup. 

I have choosen a fixed thread pool with five executors. Not sure if this is a good value or not :)

### CODE REVIEW

**Comment by bjoernhaeuser** [comment]
Not sure what the cause for the errors are, looking into them.

**Comment by bjoernhaeuser** [comment]
This could fix #461

File: src/main/java/org/jenkinsci/plugins/ghprb/GhprbTrigger.java [hunk 1]
```diff
@@ -166,6 +171,8 @@ public GhprbTrigger(String adminlist,
         this.whiteListLabels = whiteListLabels;
         setExtensions(extensions);
         configVersion = latestVersion;
+
+        this.pool = Executors.newFixedThreadPool(5);
```
Comment (chr0n1x) [inline src/main/java/org/jenkinsci/plugins/ghprb/GhprbTrigger.java]: might be better to make this configurable

Comment (bjoernhaeuser) [inline src/main/java/org/jenkinsci/plugins/ghprb/GhprbTrigger.java]: @chr0n1x can you take another look? Does this suite you better?

File: src/main/java/org/jenkinsci/plugins/ghprb/GhprbTrigger.java [hunk 2]
```diff
@@ -66,6 +68,11 @@
     @Extension
     public static final DescriptorImpl DESCRIPTOR = new DescriptorImpl();
     private static final Logger logger = Logger.getLogger(GhprbTrigger.class.getName());
+
+    private static final ExecutorService pool = Executors.newFixedThreadPool(
+        Integer.parseInt(System.getProperty("GhprbTrigger.poolSize", "5"))
```
Comment (chr0n1x) [inline src/main/java/org/jenkinsci/plugins/ghprb/GhprbTrigger.java]: is this a config that can be set in the UI or does this have to be passed in as a java arg?

Comment (bjoernhaeuser) [inline src/main/java/org/jenkinsci/plugins/ghprb/GhprbTrigger.java]: java arg. I wasnt able to make it configurable through the ui. :(

Comment (chr0n1x) [inline src/main/java/org/jenkinsci/plugins/ghprb/GhprbTrigger.java]: 😢  it would be way better if we could do it through the UI

Comment (bjoernhaeuser) [inline src/main/java/org/jenkinsci/plugins/ghprb/GhprbTrigger.java]: Agreed! Do you know how this can be achived?

Comment (chr0n1x) [inline src/main/java/org/jenkinsci/plugins/ghprb/GhprbTrigger.java]: @benpatterson halp.

I have half of that in #529 with the groovy file

**Comment by chr0n1x** [comment]
@bjoernhaeuser looks like there's a merge conflict now.
@benpatterson do you have any time to take a look at this and/or help?

**Comment by bjoernhaeuser** [comment]
Merge-Conflict solved!

**Comment by chr0n1x** [comment]
@benpatterson I think that we should still let people configure it in a range `1..Runtime.getRuntime().availableProcessors()`

I personally wouldn't want to make assumptions on what resources you do/don't have on a users' system

**Comment by benpatterson** [comment]
Ok. I may keep it simple. Or, if someone wants to tackle it, that's fine, too. Thx for pinging back.

**Comment by bjoernhaeuser** [comment]
Yeah, but has someone an idea how to make this configurable through the UI? :D

**Comment by benpatterson** [comment]
heh @bjoernhaeuser yes I think it's do-able...will take some cleaning & casting-to-int kinds of things (i.e., more work...not a ton of work, but it'll still need to get done).

**Comment by samrocketman** [comment]
Personally, I would like to see this auto-scale with `Runtime.getRuntime().availableProcessors()` but be configurable via system property for those who care.  I don't see any use in it being configurable via web UI.

File: src/main/java/org/jenkinsci/plugins/ghprb/GhprbTrigger.java (hunk 2 above)
Comment (samrocketman) [inline src/main/java/org/jenkinsci/plugins/ghprb/GhprbTrigger.java]: I recommend:
```java
import jenkins.util.SystemProperties;
...
SystemProperties.getInteger(getClass().getName() + ".poolSize", "5")
```

There's precedent in other parts of the code base like https://github.com/jenkinsci/jenkins/blob/15e69874190ce56e228b823aa1584b8497fc673b/core/src/main/java/hudson/slaves/NodeProvisioner.java#L793

**Comment by chr0n1x** [comment]
@samrocketman the only reason why I mentioned having it in the UI, with `1` being the default & system CPUs as the max, is because I don't think it's a good idea to make assumptions as to what is a "smart" use of a users' system resources. Sure, we can make a best guess but if the configuration is "hard" to change it makes the plugin a bit more inaccessible.

Just trying to keep user-empathy in mind, although I might be over-thinking it.

**Comment by samrocketman** [comment]
> Just trying to keep user-empathy in mind, although I might be over-thinking it.

It's not that I don't have empathy for users or and I'm not disregarding your feedback.  Contrary I'm sympathizing as a user myself, users want Jenkins to perform well out of the box without the need for them to heavily tune.  Detecting the number of processors at runtime is definitely reasonable and a known good practice for threaded programming.

Essentially, what I'm taking into account with my feedback is also user experience.  As a user of Jenkins, Jenkins has many checkbox items.  Adding more options for things like number of threads GhprbTrigger should use is typically a very advanced user desire.  It's better to detect resources and make a decision for the user (but fall back to allowing the user to override our decision via system property).

I would also like to point out that there's precedence for my feedback in Jenkins core as well.  [Check out `Jenkins.class` where Jenkins automatically detects the number of cores for it's own runtime](https://github.com/jenkinsci/jenkins/blob/c3b1bff6477e05050a40c0e20816bea9cf396534/core/src/main/java/jenkins/model/Jenkins.java#L4900).

I warmly give this feedback as a Jenkins user myself.  I maintain a Jenkins instance with over 3000 jobs and 1000 users.  I feel the feedback I've given is relevant to both large and small Jenkins instances which is why I want to take the time to explain to you the thoughts behind my feedback.

I care about this PR because it delays my Jenkins startup time by _more than 15 minutes_.  I would like to see it account for practices seen elsewhere in the Jenkins code base which is why I've provided links to other code justifying my feedback.

**Comment by chr0n1x** [comment]
@samrocketman completely fair & valid opinion, I was just giving my two cents is all. Like you I'm managing a master with similar numbers (thousands of jobs & users) and am coming from a perspective where I would like to tweak/tune the config multiple times depending on type of instance the master is running on via UI.

**Comment by benpatterson** [comment]
My concern is more along the lines of how many configuration bits are in this plugin, and trying to keep it simpler. Contradicting myself though, I am picking away at moving this to the UI. Although it's a good start, I don't want to hardcode it.

If someone wants to set it to the available processors, we can merge that and worry about the UI configuration later as an enhancement. Or, wait until the UI end of it gets done. No strong preference between the two right now.

**Comment by benpatterson** [comment]
Sorry; wanted to clarify.... To put it another way, I see both sides of the argument, and perhaps focusing in "when can this get done" instead, avoiding the GUI changes will speed up getting this merged.

**Comment by chr0n1x** [comment]
@benpatterson totally ok with that, thanks for the thoughts!
And yes, I'm ok with just merging this as is 😄

**Comment by samrocketman** [comment]
Does the thread pool created continue to be used post-startup during normal operation?  It was my understanding that this only affects GhprbTrigger on Jenkins startup.

**Comment by benpatterson** [comment]
Thanks for the chatter folks. Merging as-is. I'm planning on following-on with minor enhancements, but in case I can't get to that soon, the main value of getting this merge will be realized now. :dancer:

**Comment by benpatterson** [comment]
Released v 1.39.0; it'll appear in Cloudbees infra in ~24 hrs.

**Comment by samrocketman** [comment]
I'll open a PR based on my feedback since it wasn't updated.  I'll link back to this PR for discussion purposes.

**Comment by chr0n1x** [comment]
@benpatterson @samrocketman @bjoernhaeuser yaaay!!! Awesome stuff guys, thank you all 😄

**Pull request merged**